libtextsecure/message_receiver.js

```diff
@@ -120,6 +114,28 @@ MessageReceiver.prototype.extend({
     //   all cached envelopes are processed.
     this.incoming = [this.pending];
   },
+  async startLocalServer() {
+    let myLokiIp;
+    let myServerPort;
+    try {
+      myLokiIp = await window.lokiSnodeAPI.getMyLokiIp();
+    } catch (e) {
+      window.log.warn(
+        'Failed to get my loki address to bind server to, will retry in 30 seconds'
+      );
+      setTimeout(this.startLocalServer.bind(this), 30 * 1000);
+    }
+    try {
+      myServerPort = await localLokiServer.start(localServerPort, myLokiIp);
+    } catch (e) {
+      window.log.warn('Failed to start local loki server, will retry in 30 seconds');
+      setTimeout(this.startLocalServer.bind(this), 30 * 1000);
```

Comment:
You should bail after setting those timeouts :)

Also small suggestion, let's make the code flow more simple by moving the success or failure action to perform at the end:
Add a `isListening()` method in our `LocalLokiServer` class, that returns `this.server.listening`.
Then back here, basically in each `catch(e)` just prints the appropriate warning.
And only try to start the localLokiServer if we got a LokiIP first.
Then handle the success or failure.
Pseudo code:
```js
try {
  ... // get loki ip
} catch(e) {
  .. // print error
}
if (myLokiP) {
  try {
    .. // start server
  } catch(e) {
    .. // print error
  }
}
if (localLokiServer.isListening()) {
  ... // everything ok
} else {
 setTimeout(...); // retry later
}
```

Reply:
Actually I would prefer to ensure that different exception classes are thrown between `getMyLokiIp` and `LocalLokiServer.start`.
That way you could merge the two try/catches and print the correct error message by checking the error class.
```js
try {
  ... // get loki ip
  ... // start server
} catch(e) {
  .. // print error in function of e's class
}

if (localLokiServer.isListening()) {
  ... // everything ok
} else {
 setTimeout(...); // retry later
}
```

Reply:
How about this?
```js
try {
  myLokiIp = await window.lokiSnodeAPI.getMyLokiIp();
  myServerPort = await localLokiServer.start(localServerPort, myLokiIp);
  window.log.info(`Local Server started at ${myLokiIp}:${myServerPort}`);
  libloki.api.broadcastOnlineStatus();
  localLokiServer.on('message', this.handleP2pMessage.bind(this));
} catch (e) {
  if (e instanceof textsecure.LokiIpError) {
    window.log.warn(
      'Failed to get my loki address to bind server to, will retry in 30 seconds'
    );
  } else {
    window.log.warn('Failed to start local loki server, will retry in 30 seconds');
  }
  setTimeout(this.startLocalServer.bind(this), 30 * 1000);
}
```

Reply:
Yup! Even better.